modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/UndoRedo.js

```diff
@@ -0,0 +1,214 @@
+			);
+		}
+		setState({
+			defaultLanguageId: nextStep.defaultLanguageId,
+			history,
+			selectedLanguageId: nextStep.selectedLanguageId,
+			step: newStep,
+		});
+	};
+
+	const handleAutoSave = useCallback(
+		({fieldName}) => {
+			const defaultLanguageIdInput = document.getElementById(
+				`${namespace}defaultLanguageId`
+			);
+
+			const descriptionInputComponent = Liferay.componentReady(
+				`${namespace}${META_FIELD_NAMES.description}`
+			);
+
+			const titleInputComponent = Liferay.componentReady(
+				`${namespace}${META_FIELD_NAMES.title}`
+			);
+
+			const friendlyURLInputComponent = Liferay.componentReady(
+				`${namespace}${META_FIELD_NAMES.friendlyURL}`
+			);
+
+			const selectedLanguageIdInput = document.getElementById(
+				`${namespace}languageId`
+			);
+
+			Promise.all([
+				descriptionInputComponent,
+				titleInputComponent,
+				friendlyURLInputComponent,
+			]).then(
+				([
+					descriptionInputComponent,
+					titleInputComponent,
+					friendlyURLInputComponent,
+				]) => {
+					const newHistory = {
+						defaultLanguageId: defaultLanguageIdInput.value,
+						descriptionInputComponent: descriptionInputComponent.getValue(
+							selectedLanguageId
+						),
+						friendlyURLInputComponent: friendlyURLInputComponent.getValue(
+							selectedLanguageId
+						),
+						name: fieldName,
+						selectedLanguageId: selectedLanguageIdInput.value,
+						titleInputComponent: titleInputComponent.getValue(
+							selectedLanguageId
+						),
+					};
+
+					setState({
+						defaultLanguageId: defaultLanguageIdInput.value,
+						history: [...history.slice(0, step + 1), newHistory],
+						selectedLanguageId: selectedLanguageIdInput.value,
+						step: step + 1,
+					});
+				}
+			);
+		},
+		[history, namespace, selectedLanguageId, step]
+	);
+
+	const localeChangeHandler = useCallback(
+		(event) => {
+			const selectedLanguageId = event.item.getAttribute('data-value');
+
+			const selectedLanguageIdInput = document.getElementById(
+				`${namespace}languageId`
+			);
+
+			selectedLanguageIdInput.value = selectedLanguageId;
+
+			Liferay.fire('autoSave', {fieldName: 'Locale Change'});
+		},
+		[namespace]
+	);
+
+	useEffect(() => {
+		Liferay.after('inputLocalized:localeChanged', localeChangeHandler);
+
+		return () => {
+			Liferay.detach('inputLocalized:localeChanged', localeChangeHandler);
+		};
+	}, [localeChangeHandler]);
+
+	useEffect(() => {
+		Liferay.on('autoSave', handleAutoSave);
+
+		return () => {
+			Liferay.detach('autoSave', handleAutoSave);
+		};
+	}, [handleAutoSave]);
+
+	useEffect(() => {
+		setTimeout(() => {
+			Liferay.fire('autoSave', {fieldName: 'Reset'});
+		}, 1000);
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, []);
+
+	return (
+		<div>
+			<ClayButtonWithIcon
+				aria-label={Liferay.Language.get('undo')}
+				className="btn-monospaced"
+				disabled={step <= 0}
+				displayType="secondary"
+				onClick={() => {
+					Liferay.fire('undo');
```

Comment:
We are not listening these events in other components, so no need to fire this (nor the 'redo' event)

Reply:
We are listening in data-engine. Restoring data-engine to a previous or future state requires of firing this event